File: frontend/src/lib/api/course-templates.ts
/**
 * Uploaded course template documents (DOCX/Markdown, normalized to
 * Markdown server-side). Separate from lib/api/templates.ts, which manages
 * the admin JSON pipeline-schema templates - a different feature that
 * happens to share the word "template".
 */

import { request, requestForm } from "./client";

export type CourseTemplateType = "technical" | "non_technical";

export interface CourseTemplateDocumentSummary {
  id: string;
  name: string;
  template_type: CourseTemplateType;
  description: string;
  source_format: "docx" | "md";
  content_format: "markdown";
  created_by: string | null;
  created_at: string;
  updated_at: string;
}

export interface CourseTemplateDocumentDetail extends CourseTemplateDocumentSummary {
  markdown_content: string;
}

export async function listCourseTemplateDocuments(): Promise<CourseTemplateDocumentSummary[]> {
  const response = await request<{ templates: CourseTemplateDocumentSummary[] }>(
    "/api/course-template-documents",
  );
  return response.templates;
}

export async function getCourseTemplateDocument(id: string): Promise<CourseTemplateDocumentDetail> {
  return request<CourseTemplateDocumentDetail>(`/api/course-template-documents/${id}`);
}

export async function uploadCourseTemplateDocument(input: {
  name: string;
  templateType: CourseTemplateType;
  description: string;
  file: File;
}): Promise<CourseTemplateDocumentDetail> {
  const form = new FormData();
  form.set("name", input.name);
  form.set("template_type", input.templateType);
  form.set("description", input.description);
  form.set("file", input.file);
  return requestForm<CourseTemplateDocumentDetail>("/api/course-template-documents", form);
}

export interface TemplateTypeClassification {
  template_type: CourseTemplateType;
  reasoning: string;
}

/** Backs the hardcoded "Default Template" option: given the course title/
 * audience entered so far, asks the backend to decide Technical vs
 * Non-Technical rather than making the user pick. */
export function classifyTemplateType(input: {
  courseTitle: string;
  targetAudience: string;
}): Promise<TemplateTypeClassification> {
  return request<TemplateTypeClassification>("/api/course-template-documents/classify-type", {
    method: "POST",
    body: { course_title: input.courseTitle, target_audience: input.targetAudience },
  });
}

File: backend/app/api/course_templates.py
"""Uploaded course template documents - any authenticated user can upload,
browse or select. Separate from /admin/templates (the JSONB pipeline-schema
templates in api/templates.py, still admin-only) - these are prose reference
documents (DOCX or Markdown, normalized to Markdown) offered on Create
Course, not wired into generation yet.
"""

from __future__ import annotations

from uuid import UUID

from fastapi import APIRouter, Depends, File, Form, UploadFile
from pydantic import ValidationError

from app.api.dependencies import get_current_user
from app.core.errors import ValidationFailedError
from app.db.models import User
from app.schemas.course_template_document import (
    CourseTemplateDocumentDetail,
    CourseTemplateDocumentListResponse,
    CourseTemplateDocumentUploadForm,
    TemplateTypeClassification,
    TemplateTypeClassifyRequest,
)
from app.services.course_template_document_service import (
    CourseTemplateDocumentService,
    get_course_template_document_service,
)

router = APIRouter(prefix="/api/course-template-documents", tags=["course-template-documents"])


def _service() -> CourseTemplateDocumentService:
    return get_course_template_document_service()


@router.get("", response_model=CourseTemplateDocumentListResponse)
async def list_course_template_documents(
    service: CourseTemplateDocumentService = Depends(_service),
    current_user: User = Depends(get_current_user),
) -> CourseTemplateDocumentListResponse:
    return CourseTemplateDocumentListResponse(templates=await service.list_active())


@router.get("/{template_id}", response_model=CourseTemplateDocumentDetail)
async def get_course_template_document(
    template_id: UUID,
    service: CourseTemplateDocumentService = Depends(_service),
    current_user: User = Depends(get_current_user),
) -> CourseTemplateDocumentDetail:
    return await service.get(template_id)


@router.post("/classify-type", response_model=TemplateTypeClassification)
async def classify_template_type(
    request: TemplateTypeClassifyRequest,
    service: CourseTemplateDocumentService = Depends(_service),
    current_user: User = Depends(get_current_user),
) -> TemplateTypeClassification:
    """Backs the hardcoded "Default Template" option: given what's been
    entered on Create Course so far, decide technical vs non_technical."""
    return await service.classify_type(request.course_title, request.target_audience)


@router.post("", status_code=201, response_model=CourseTemplateDocumentDetail)
async def upload_course_template_document(
    name: str = Form(...),
    template_type: str = Form(...),
    description: str = Form(""),
    file: UploadFile = File(...),
    service: CourseTemplateDocumentService = Depends(_service),
    current_user: User = Depends(get_current_user),
) -> CourseTemplateDocumentDetail:
    try:
        form = CourseTemplateDocumentUploadForm(
            name=name, template_type=template_type, description=description
        )
    except ValidationError as exc:
        readable = "; ".join(f"{'.'.join(str(p) for p in err['loc'])}: {err['msg']}" for err in exc.errors())
        raise ValidationFailedError(readable) from exc
    return await service.upload(
        form,
        file,
        created_by=current_user.id,
        created_by_email=current_user.email,
    )
